# Exercise Set 13: Network formation


In this Exercise Set 13 we will investigate network formation among high school pupils. 

## Part 1: Network formation


Load the data using the script below. Read a bit about the dataset [here](http://www.sociopatterns.org/datasets/high-school-contact-and-friendship-networks/) to get an understanding of what is in each variable. 

The script gives you two dataframes to work with: 
 > `el`, which is an edge-list 
 >
 > `ind` which contains individual characteristics

In [1]:
import networkx as nx
import numpy as np
import pandas as pd

url_base = 'http://www.sociopatterns.org/wp-content/uploads/2015/'

# edgelist
url_el = url_base + '07/High-School_data_2013.csv.gz'
col_names_el = ['timestamp', 'u1', 'u2', 'class1', 'class2']
el = pd.read_csv(url_el, header=None, names=col_names_el, delimiter=' ')

# individual characteristics
url_ind = url_base + '09/metadata_2013.txt'
col_names_ind = ['u', 'class', 'gender']
ind = pd.read_csv(url_ind, header=None, names=col_names_ind, delimiter='\t')\
            .set_index('u')

# remove observation with missing gender
has_gender = ind[ind.gender!='Unknown'].index

# DataFrames
ind = ind.loc[has_gender].copy()
el = el[el.u1.isin(has_gender) &  el.u2.isin(has_gender)].copy()

In [2]:
print(el.head())
print(ind.head())

    timestamp   u1   u2 class1 class2
0  1385982020  454  640     MP     MP
1  1385982020    1  939  2BIO3  2BIO3
2  1385982020  185  258    PC*    PC*
3  1385982020   55  170  2BIO3  2BIO3
4  1385982020    9  453     PC     PC
     class gender
u                
650  2BIO1      F
498  2BIO1      F
627  2BIO1      F
857  2BIO1      F
487  2BIO1      F


In [3]:
el2=el

> **Ex. 13.1.1**: Describe the edgelist columns content. Parse the timestamp. What is the resolution of meetings? Use the parsed timestamp to count the meetings by hour in local time.

[Answer to ex. 13.1.1 here]
The edgelist (el) consists of 4 columns, Timestamp, u1, u2, class1, class2 
* Timestamp is the time interval where two students was in contact and is measured in seconds and given in UNIX time (seconds since Jan 01 1970. (UTC)). 
* u1 is person 1 which have had contact with person 2 (u2). 
* Class1 and class2 is the classes for person 1 and 2, respectively. 

The timestamp is given in unix time wich is seconds since Jan 01 1970 (UTC). In this case the timestamp is given as a time interval of 20 seconds. Therefore, many of the timestamps have the same value. 

The first timestamp (1385982020) is equal to 12/02/2013, 11:00am (UTC)
and the last (1386345580) is equal to 12/06/2013,  3:59pm (UTC). Which is equal to the 5 days the data collection was taking place.


In [4]:
el2

timestamp    u1    u2 class1 class2
0       1385982020   454   640     MP     MP
1       1385982020     1   939  2BIO3  2BIO3
2       1385982020   185   258    PC*    PC*
3       1385982020    55   170  2BIO3  2BIO3
4       1385982020     9   453     PC     PC
5       1385982020     9    45     PC     PC
6       1385982020    14   190    PC*    PC*
7       1385982020   400   637  2BIO1  2BIO1
8       1385982020   255   275  2BIO3  2BIO3
9       1385982020   176   533    PC*    PC*
10      1385982020   116   533    PC*    PC*
11      1385982020   151   866     MP     MP
12      1385982020   280   484   MP*1   MP*1
14      1385982020    54   866     MP     MP
15      1385982020   364   374  2BIO3  2BIO3
16      1385982020   295   441  2BIO2  2BIO2
17      1385982020   101   425  2BIO3  2BIO3
18      1385982020    47   241   PSI*   PSI*
19      1385982020   116   176    PC*    PC*
20      1385982020   179   202  2BIO3  2BIO3
21      1385982020    63   179  2BIO3  2BIO3
22      1385982020   564   577     MP     MP
23      1385982020   265   494  2BIO3  2BIO3
24      1385982020   443   637  2BIO1  2BIO1
26      1385982020   222   843    PC*    PC*
27      1385982020   205   894  2BIO3  2BIO3
28      1385982020  1359  1383   MP*2   MP*2
29      1385982020   376   638   MP*1   MP*1
30      1385982020  1238  1260   MP*2   MP*2
31      1385982020   487   984  2BIO1  2BIO1
...            ...   ...   ...    ...    ...
188477  1386345460   120   311     PC     PC
188478  1386345460   120   285     PC     PC
188479  1386345480   272   939  2BIO3  2BIO3
188480  1386345480   176   258    PC*    PC*
188481  1386345480   120   311     PC     PC
188482  1386345480   120   285     PC     PC
188483  1386345500   272   939  2BIO3  2BIO3
188484  1386345500   176   258    PC*    PC*
188485  1386345500   274   484   MP*1   MP*1
188486  1386345500   311   453     PC     PC
188487  1386345500   120   311     PC     PC
188488  1386345500   285   311     PC     PC
188489  1386345500   120   285     PC     PC
188490  1386345520   272   939  2BIO3  2BIO3
188491  1386345520   176   258    PC*    PC*
188492  1386345520   525   687     MP     MP
188494  1386345520   285   311     PC     PC
188495  1386345520   120   285     PC     PC
188496  1386345540   272   939  2BIO3  2BIO3
188497  1386345540   219   687     MP     MP
188498  1386345540   120   285     PC     PC
188499  1386345560   280   513   MP*1   MP*1
188500  1386345560   200   311     PC     PC
188501  1386345560   272   939  2BIO3  2BIO3
188502  1386345560   311   496     PC     PC
188503  1386345560   120   285     PC     PC
188504  1386345580    61   160  2BIO2  2BIO2
188505  1386345580   272   939  2BIO3  2BIO3
188506  1386345580   311   496     PC     PC
188507  1386345580   120   285     PC     PC

[181649 rows x 5 columns]

In [5]:
el2['datetime']=pd.to_datetime(el2["timestamp"], unit='s')
meetings=el2.groupby([pd.Grouper(key='datetime',freq='H')]).size().reset_index(name='count')


In [6]:
meetings.query('count != 0')

datetime  count
0   2013-12-02 11:00:00   5556
1   2013-12-02 12:00:00   4259
2   2013-12-02 13:00:00   6617
3   2013-12-02 14:00:00   5715
4   2013-12-02 15:00:00   5972
20  2013-12-03 07:00:00   6048
21  2013-12-03 08:00:00   5286
22  2013-12-03 09:00:00   7104
23  2013-12-03 10:00:00   5096
24  2013-12-03 11:00:00   4675
25  2013-12-03 12:00:00   4193
26  2013-12-03 13:00:00   5172
27  2013-12-03 14:00:00   3772
28  2013-12-03 15:00:00   4316
44  2013-12-04 07:00:00   5100
45  2013-12-04 08:00:00   6218
46  2013-12-04 09:00:00   7309
47  2013-12-04 10:00:00   4013
48  2013-12-04 11:00:00   3998
49  2013-12-04 12:00:00   4555
50  2013-12-04 13:00:00   3109
51  2013-12-04 14:00:00   2567
52  2013-12-04 15:00:00   2117
68  2013-12-05 07:00:00   4603
69  2013-12-05 08:00:00   4851
70  2013-12-05 09:00:00   6146
71  2013-12-05 10:00:00   4230
72  2013-12-05 11:00:00   3063
73  2013-12-05 12:00:00   3039
74  2013-12-05 13:00:00   3680
75  2013-12-05 14:00:00   3461
76  2013-12-05 15:00:00   2595
92  2013-12-06 07:00:00   3877
93  2013-12-06 08:00:00   4872
94  2013-12-06 09:00:00   6898
95  2013-12-06 10:00:00   5051
96  2013-12-06 11:00:00   4106
97  2013-12-06 12:00:00   3247
98  2013-12-06 13:00:00   1785
99  2013-12-06 14:00:00   2026
100 2013-12-06 15:00:00   1352

> **Ex. 13.1.2**: Count the number of meetings for each edge and save this as a DataFrame called `el_agg`. Filter out edges with less than 5 minutes of meetings. Attach the gender and class of both nodes.

In [7]:
# [Answer to ex. 13.1.2 here]
el_agg=el2.groupby(['u1','u2']).size().reset_index(name='count')

In [8]:
el_agg=el_agg.query('count > 5 ').reset_index()

In [9]:
el_agg=pd.merge(el_agg, ind, left_on='u1', right_on='u')
el_agg=pd.merge(el_agg, ind, left_on='u2', right_on='u')

In [24]:
el_agg=el_agg.sort_values('u1').reset_index()

In [26]:
el_agg=el_agg.rename(columns={"class_x": "class_u1", "gender_x": "gender_u1","class_y": "class_u2", "gender_y": "gender_u2"})

In [47]:
el_agg.head(10)

index  u1   u2  count class_u1 gender_u1 class_u2 gender_u2
0      0   1   55      8    2BIO3         M    2BIO3         F
1     21   1  939     85    2BIO3         M    2BIO3         M
2     10   1  205     47    2BIO3         M    2BIO3         M
3     16   1  778     10    2BIO3         M    2BIO3         F
4      7   1  196     38    2BIO3         M    2BIO3         M
5      5   1  170      8    2BIO3         M    2BIO3         F
6     13   1  494    123    2BIO3         M    2BIO3         M
7      4   1  117     18    2BIO3         M    2BIO3         M
8     65   3  653     97    2BIO2         M    2BIO3         M
9     24   3   28     27    2BIO2         M    2BIO2         M

> **Ex. 13.1.3**: Answer question in the function `fraction_triangles` below. Explain how `fraction_triangles` is related to  computing the clustering coefficient (using `nx.average_clustering`).
>
>> *Hint:* The following code does the same thing as `fraction_triangles`, but at a scale where you can understand what's going on. If you have a hard time understanding the code in the function you can try to play around with this simpler example
>>
>> ```python
>> import networkx as nx 
>>
>> A  = np.array(
>>     [[0, 1, 1, 0],
>>      [1, 0, 1, 0],
>>      [1, 1, 0, 1],
>>      [0, 0, 1, 0]]
>> )
>>
>> G = nx.from_numpy_array(A)
>> nx.draw(G,with_labels=True)
>>
>> def nth(A, n):
>>     A_ = A.copy()    
>>     for _ in range(1,n):
>>         A = A.dot(A_)
>>     return A
>>
>> a_t = nth(A,3).diagonal().sum()/6
>> n = len(A[:,0])
>> p_t = binom(n, 3)
>> ```


In [31]:
def make_net(el_, nodes):
    '''
    Convert edgelist to networkx graph which is 
    binary and undirected.
    
    Parameters
    ----------
    el_ : DataFrame
        Table containing an edgelist with columns 
        `u1` and `u2` which are the nodes in the edge.
        
    nodes : array-like
        1d array containing the node identities.
    '''    
    
    nx_input = el_, 'u1', 'u2', 'count', nx.Graph()
    g = nx.from_pandas_edgelist(*nx_input)
    g.add_nodes_from(nodes)
    return g

In [32]:
from scipy.special import binom

def fraction_triangles(el_, nodes):
    '''
    Compute fraction of actual triangles out 
    of the potential triangles.
    
    Parameters
    ----------
    el_ : DataFrame
        Table containing an edgelist with columns 
        `u1` and `u2` which are the nodes in the edge.
        
    nodes : array-like
        1d array containing the node identities.
    '''
    
    g = make_net(el_, nodes)
    
    #Q.1: what is `A`?: the adjacency matrix which is symmetric and binary
    #Q.2: what does `A**3` do? compute the number of paths between two nodes
    #Q.3: what is diagonal of A_t? the number of actual paths of length 3, 
    # i.e. triangles, which include the person. these are called cycles
    # because they start and end at the same person
    
    # count actual triangles    
    A = nx.to_scipy_sparse_matrix(g)
    A_t = A**3
    a_t = A_t.diagonal().sum()/6
    
    #Q.4: what does `binom(n,3)` compute? the number of triangles including the person
    
    # count potential triangles
    n = len(g.nodes())
    p_t = binom(n, 3)
        
    return a_t/p_t

Q.1: what is `A`?: the adjacency matrix which is symmetric and binary
Q.2: what does `A**3` do? compute the number of paths between two nodes
Q.3: what is diagonal of A_t? the number of actual paths of length 3,  i.e. triangles, which include the person. these are called cycles because they start and end at the same person
Q.4: what does `binom(n,3)` compute? the number of triangles including the person


The first function make_net is creating a network from the edgelist and returns a network graph. The second function is creating the fraction of actual triangles out of potential triangles. (triangles between edges). Spicy_sparse_matirx is creating the adjacent matrix.

> **Ex. 13.1.4**: Apply the function `fraction_triangles` to `el_agg` and print the triangle fraction in the network. Next remove all edges that go between classes. Compute triangle fraction within each class and store it. Compute the mean within class triangles and bootstrap the standard error of the mean. Comment on the output.
>
>> *Hint:* To bootstrap an estimate draw $k>>0$ samples with replacement from the data. Compute the estimate on each of these samples and average them in the end to get the bootstrapped estimate. 

In [39]:
# [Answer to ex. 13.1.4 here]
fraction_triangles(el_agg,el_agg['u1'])

0.0009860959537600295

In [53]:
el_agg2=el_agg.drop(el_agg[el_agg['class_u1'] != el_agg['class_u2']].index)

In [60]:
#Unique classes in the data.
class_values=[]

for class_ in el_agg2['class_u1'].unique():
    x=el_agg2.loc[el_agg2['class_u1']==class_]
    y=x['u1']
    z=fraction_triangles(x,y)
    class_values.append(z)

In [77]:
class_values1= pd.DataFrame(class_values)
classes=pd.DataFrame(el_agg2['class_u1'].unique())

Class_frac_tri=pd.merge(classes,class_values1, left_index=True, right_index=True).rename(columns={"0_x": "Class", "0_y": "fraction_triangle"})
Class_frac_tri

Class  fraction_triangle
0  2BIO3           0.105466
1  2BIO2           0.078996
2   PSI*           0.041606
3     PC           0.086756
4    PC*           0.078507
5   MP*1           0.031441
6     MP           0.107827
7  2BIO1           0.036835
8   MP*2           0.077762

Recall from class that we can define the following measures of homophily. We define **homophily index** inspired by [Currarini et al. (2009)](https://doi.org/10.2139/ssrn.1021650):
- share of edges that are same type: $H = \frac{s}{s+d}$
- possible range [0,1]


We define **baseline homophily** as: 
- We count fraction of potential edges in population of nodes which are same type:

\begin{equation}B=\frac{\sum_t\#potential(n_t)}{\#potential(n)}, \qquad \#potential(k)=\frac{k\cdot(k-1)}{2}\end{equation}

- Interpretation: Expected homophily from random link formation.     

We define **inbreeding homophily** as:      

\begin{equation}IH=\frac{H-B}{1-B}\end{equation}


> **Ex. 13.1.5**: Compute the inbreeding homophily for each class. Use the class measures to compute the mean. Use a bootstrap to compute whether there is inbreeding homophily.

In [138]:
# [Answer to ex. 13.1.5 here]
def IH(df):
    
    "s is the 𝑠 : number of edges with same type"
    "𝑑 is the number of edges with different type"
    
    fgt=df[df['class1']==df['class2']].count()
    s=fgt['class1']
    jio=df[df["class1"]!=df["class2"]].count()
    d=jio['class1']
    
    'H: share of edges that are same type'
    H=s/(s+d)
    
    'We count fraction of potential edges in population of nodes which are same type:'
    'This is done by the function below'
    def potential(k):
        return (k*(k-1))/2
    
    'potential given where the classes are the same'
    B=potential(s)/potential(len(df))
    
    return (H-B)/(1-B)

In [139]:
class_IH=[]

for class_ in el2['class1'].unique():
    x=el2.loc[el2['class1']==class_]
    z=IH(x)
    class_IH.append(z)

C:\Python\Anaconda3\lib\site-packages\ipykernel_launcher.py:23: RuntimeWarning: invalid value encountered in double_scalars


In [145]:
d_class_IH=pd.DataFrame(class_IH)
print('mean of classes' ,d_class_IH.mean())

mean of classes 0    0.47614
dtype: float64


> **Ex. 13.1.6** (BONUS): Describe what an unsupported edge is. Construct a test of whether there is a preference for forming  triangles within same gender than across.
>
>> *Hint:*  You can find inspiration in the approach of [Chandrasekhar, Jackson (2018)](https://web.stanford.edu/~arungc/CJ_sugm.pdf) pp. 31-35. They construct an almost identical test for triangle formation across castes in Indian villages.

In [15]:
# [Answer to ex. 13.1.6 here]